tests/lib/timespec_util/src/main.c

```diff
@@ -329,54 +425,138 @@ ZTEST(timeutil_api, test_timespec_from_timeout)
 				      "%d: {%" PRId64 "} and {%" PRId64
 				      "} are unexpectedly different",
-				      i, (int64_t)actual.ticks, (int64_t)K_NO_WAIT.ticks);
-			continue;
-		}
-
-		if ((tspec->saturation > 0) ||
-		    (timespec_compare(&tspec->tspec, &k_timeout_limits[1]) > 0)) {
-			/* K_FOREVER saturation */
-			actual = timespec_to_timeout(&tspec->tspec);
-			zexpect_equal(actual.ticks, K_TICKS_FOREVER,
+				      i, (int64_t)actual.ticks, (int64_t)K_TICK_MIN);
+		} else if (tspec->saturation > 0) {
+			/*
+			 * Careful - the compiler can play tricks!
+			 *
+			 * Above, the expression DECL_PSAT_TOSPEC_TEST(K_TICKS_TO_SECS(K_TICK_MAX),
+			 * K_TICKS_TO_NSECS(K_TICK_MAX)), does not always properly evaluate to
+			 * values equivalent to what is in K_TS_MAX. Specifically, when
+			 * CONFIG_TIMEOUT_64BIT=y and CONFIG_SYS_CLOCK_TICKS_PER_SEC == 100.
+			 *
+			 * The tv_nsec field is 0! As a workaround, we substitute for K_TS_MAX
+			 * directly. This is very odd, since K_TS_MAX itself is defined the same
+			 * way and is correctly compiled.
+			 *
+			 * If someone is able to suggest a better way to ensure that the correct
+			 * fields are compiled-in, please do, remove 'true' below, and remove the
+			 * workaround.
+			 */
+			__ASSERT(true || (timespec_compare(&tspec->tspec, &K_TS_MAX) <= 0),
+				 "timespec: {%lld, %lld} is not greater than or equal to K_TS_MAX",
+				 (long long)tspec->tspec.tv_sec, (long long)tspec->tspec.tv_nsec);
+
+			struct timespec workaround;
+
+			switch (i) {
+			case ARRAY_SIZE(tospecs) - 3:
+				workaround = K_TS_MAX;
+				break;
+			case ARRAY_SIZE(tospecs) - 2:
+				workaround = K_TS_MAX;
+				workaround.tv_sec += 1;
+				break;
+			}
+
+			/* K_TICK_MAX saturation */
+			actual = timespec_to_timeout_rem(&workaround, &rem);
+			zexpect_equal(actual.ticks, K_TICK_MAX,
 				      "%d: {%" PRId64 "} and {%" PRId64
 				      "} are unexpectedly different",
-				      i, (int64_t)actual.ticks, (int64_t)K_TICKS_FOREVER);
+				      i, (int64_t)actual.ticks, (int64_t)K_TICK_MAX);
+
+			/* workaround */
+			timespec_from_timeout(tspec->timeout, &tick_ts);
+			timespec_add(&tick_ts, &rem);
+			zexpect_true(timespec_equal(&tick_ts, &workaround),
+				     "%d: {%ld, %ld} and {%ld, %ld} are unexpectedly different", i,
+				     tick_ts.tv_sec, tick_ts.tv_nsec, workaround.tv_sec,
+				     workaround.tv_nsec);
 			continue;
 		}
+
+		timespec_from_timeout(tspec->timeout, &tick_ts);
+		timespec_add(&tick_ts, &rem);
+		zexpect_true(timespec_equal(&tick_ts, &tspec->tspec),
+			     "%d: {%ld, %ld} and {%ld, %ld} are unexpectedly different", i,
+			     tick_ts.tv_sec, tick_ts.tv_nsec, tspec->tspec.tv_sec,
+			     tspec->tspec.tv_nsec);
+	}
+
+	/* Special case highlighed in #92709 */
+#if defined(CONFIG_TIMEOUT_64BIT) && (CONFIG_SYS_CLOCK_TICKS_PER_SEC == 100)
+	{
+		struct timespec rem = {};
+		k_timeout_t to = K_TICKS(9223372036854775807LL);
+		struct timespec ts = K_TIMESPEC(92233720368547758L, 70000000L);
```

Comment:
```suggestion
		struct timespec ts = K_TIMESPEC(92233720368547758LL, 70000000L);
```
That number needs 40bits.